app/src/main/java/com/eveningoutpost/dexdrip/Services/broadcastservice/BroadcastService.java

```diff
@@ -495,6 +527,9 @@ protected Bundle prepareBgBundle(BroadcastModel broadcastModel) {
                 bundle.putLong("treatment.timeStamp", treatment.timestamp);
             }
 
+            predictedIOB = "";
```

Comment:
IMHO this always sets the variables to an empty string which does not change till line 569. Maybe you should initialize the strings to `""` in line 84/85?

Reply:
This is a place where the variable resets in case if the iob data would not be updated in BgGraphBuilder.

Reply:
But in case if settings.isDisplayGraph() is false, the predictedIOB and predictedBWP always would be empty. Maybe it would be better to transfer iob only if graph is enabled.

Reply:
The only location where `predictedIOB` and `predictedBWP` are set to likely non-empty value is in the `statusReceiver` `onReceive` event. The next use is `bundle.putString("predict.IOB", predictedIOB);` but right before the if condition these two variables will be emptied. Hence, en empty string is always passed to `bundle.putString`. That's my main concern. Did I oversee anything?

Reply:
Yes. the variables are filled via statusReceiver by receiving the broadcast messages.  And those broadcast messages can  triggers by bgGraphBuilder.defaultLines(false)  function which calls addBgReadingValues(simple) and the last one function calls  Home.updateStatusLine("iob", df.format(iob.iob))  . So when the code reaches to the bundle.putString , the **iob** and **wpb** values would be filled with the valid data.

Reply:
Okay, many of the functions are annotated with `synchronized` but `sendBroadcast(intent)` broadcasts the given intent to all interested BroadcastReceivers [asynchronously](https://developer.android.com/reference/android/content/Context#sendBroadcast(android.content.Intent). How can you be sure that all operations are finished until the `bundle.putString` lines are reached?

Reply:
You are absolutely right that there is no guarantee that broadcast msg will be received at the moment the bundle.putString lines are reached.  I thought about this and do not found any other way to make sure the predictedIOB  would be cleared.  The similar broadcast receiver for predictedIOB  is located in Home.java and it is reset to "" by onResume which is also not the best place to reset this info. But it is at least reset the value after you reopen the app.

Reply:
How about this:


```java
    private String predictedBWP = "";
    private String predictedIOB = "";

...

            public void onReceive(Context context, Intent intent) {
                final String bwp = intent.getStringExtra("bwp");
                if (bwp != null) {
                    predictedBWP = bwp;
                } else {
                    predictedBWP = "";            
                }
                final String iob = intent.getStringExtra("iob");
                if (iob != null) {
                    predictedIOB = iob;
                } else {
                    predictedIOB = "";            
                }
            }
```

and in `onBind` moving `predictedBWP = ""; predictedIOB = "";` after `bundle.putString`.

PS: I've moved out `iob` from the `bwp` else part because i thought that these values are set independently (in certain cases). Is this correct or is `iob` only set/changed if `bwp` is set/changed?

Reply:
In the `BgGraphBuilder.java `you can see that `Home.updateStatusLine("bwp", bwp_update) `will triggers almost always, while the `prediction_enabled && simulation_enabled`. 
The `Home.updateStatusLine("iob", df.format(iob.iob)) `has more complex logic. So in most cases the bwp would trigger broadcast. But this will happen until the user did not switch off the  `simulation_enabled` option.

Reply:
Ah, It looks like if use your proposed code, the iob would be always empty, because when BgGraphBuilder will send  **bwp** broadcast, it will clear the **iob** info ...

Reply:
@tolot27 as i understand this pr was stuck because the "bwp" and "iob" broadcast data which is generated in BgGraphBuilder transfers with help of sendBroadcast finction which is runs asynchronously. Maybe to resolve this problem we can change  sendBroadcast in [updateStatusLine ](https://github.com/NightscoutFoundation/xDrip/blob/master/app/src/main/java/com/eveningoutpost/dexdrip/Home.java#L690) to [sendBroadcastSync](https://developer.android.com/reference/androidx/localbroadcastmanager/content/LocalBroadcastManager#sendBroadcastSync(android.content.Intent))  which should block the execution of the thread .

Reply:
What about using the keystore data like is done with `bwp_last_insulin` in `PebbleDisplayTrend` ? I could easily add more keystore fields if this doesn't include everything needed. Then you don't need to worry about asynchronous broadcasts.

Reply:
Yep. I guess this should help to transfer data across different parts of the application.  It is enough to add bwp_update data (but maybe better to divide the current string inot two separate variables with the carb and insulin info). And a separate keystore for "iob" which is declared by this line ` Home.updateStatusLine("iob", df.format(iob.iob));`

Reply:
Yes. Now works fine, added appropriate code changes